src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -1679,22 +1679,29 @@ private boolean areDisjoint(ClassSymbol ts, ClassSymbol ss) {
                  */
                 if (!pairsSeen.add(newPair))
                     return false;
-                if (isSubtype(erasure(ts.type), erasure(ss.type))) {
-                    return false;
-                }
-                // if both are classes or both are interfaces, shortcut
-                if (ts.isInterface() == ss.isInterface() && isSubtype(erasure(ss.type), erasure(ts.type))) {
-                    return false;
-                }
-                if (ts.isInterface() && !ss.isInterface()) {
-                    /* so ts is interface but ss is a class
-                     * an interface is disjoint from a class if the class is disjoint form the interface
+                // if both are classes
+                if (!ts.isInterface() && !ss.isInterface()) {
+                    return !isSubtype(erasure(ss.type), erasure(ts.type)) && !isSubtype(erasure(ts.type), erasure(ss.type));
+                } else if (ts.isInterface() && !ss.isInterface() || !ts.isInterface() && ss.isInterface()) {
+                    /* so one is an interface and the other one is a class
                      */
-                    return areDisjoint(ss, ts);
-                }
-                // a final class that is not subtype of ss is disjoint
-                if (!ts.isInterface() && ts.isFinal()) {
-                    return true;
+                    ClassSymbol isym = ts.isInterface() ? ts : ss;
+                    ClassSymbol csym = isym == ts ? ss : ts;
+                    if (isSubtype(erasure(csym.type), erasure(isym.type))) {
+                        return false;
+                    } else {
+                        if (csym.isFinal()) {
+                            return true;
```

Comment:
Maybe add a comment e.g. `not freely extensible`

Reply:
Strike that - freely extensible is checked later. Seems to me that this `csym.isFinal()` check is subsumed by the new `isClassFreelyExtensible` predicate (and so, can be removed) ?

Reply:
Question: I find it hard to verify that this logic matches the spec exactly. That's because the spec says when under which conditions two types are disjoint, while this code seems to work the other way around -- e.g. find ways to say when two types are NOT disjoint.

Would it be possible to write the code the other way around and align it fully with the spec?

Reply:
(Note - the problems above are pre-existing, but since we're here we might as well clean up for good?)